### PR TITLE
[GH-3156][GH-3157][GH-3158] Implement distributed GeoSeries methods

### LINKED ISSUE #3156: GeoPandas: implement distributed GeoSeries.get_coordinates
Part of #2230.

Implement `GeoSeries.get_coordinates(include_z=False, ignore_index=False, index_parts=False, *, include_m=False)` with GeoPandas-compatible behavior.

Requirements:
- Execute fully in Spark using Sedona geometry expressions; do not collect geometries to the driver and do not use a Python UDF.
- Return a distributed pandas-on-Spark DataFrame with `x` and `y` columns, plus optional `z` and `m` columns.
- Preserve flattened vertex order across all supported geometry families.
- Represent missing optional ordinates consistently with GeoPandas.
- Preserve the original index by default.
- When `index_parts=True`, return a two-level index containing the original index and zero-based coordinate position.
- When `ignore_index=True`, return a new distributed sequential index.
- Match GeoPandas behavior for empty geometries and null values.

Add focused parity tests for geometry families, Z and M coordinates, mixed dimensionality, empty and null inputs, named and multi-level indexes, all index modes, and ordering. Add the public API documentation required for a new GeoPandas-compatible method.

### LINKED ISSUE #3157: GeoPandas: implement distributed GeoSeries.explode
Part of #2230.

Implement `GeoSeries.explode(ignore_index=False, index_parts=False)` with GeoPandas-compatible behavior.

Requirements:
- Execute fully in Spark using Sedona geometry expressions; do not collect geometries to the driver and do not use a Python UDF.
- Expand multipart geometries and geometry collections into their immediate component geometries while keeping simple geometries as one row.
- Preserve geometry dtype, CRS, series name, and deterministic component order.
- Preserve the original index by default.
- When `index_parts=True`, return a two-level index containing the original index and zero-based component position.
- When `ignore_index=True`, return a new distributed sequential index.
- Match GeoPandas behavior for empty geometries, empty multipart geometries or collections, and null values.

Add focused parity tests for geometry families, empty and null inputs, named and multi-level indexes, all index modes, CRS, series name, and ordering. Add the public API documentation required for a new GeoPandas-compatible method.

### LINKED ISSUE #3158: GeoPandas: implement distributed GeoSeries.geom_equals_exact
Part of #2230.

Implement `GeoSeries.geom_equals_exact(other, tolerance, align=None)` with GeoPandas-compatible behavior.

Requirements:
- Execute fully in Spark using a native Sedona predicate; do not collect geometries to the driver and do not use a Python UDF.
- Add a native exact-equality predicate with tolerance when the existing public predicates cannot express the required semantics cleanly.
- Support scalar geometry and GeoSeries operands.
- Match GeoPandas index alignment behavior for `align=None`, `align=True`, and `align=False`.
- Return a distributed boolean Series with the source index and GeoPandas-compatible null behavior.
- Match exact structural equality semantics, including coordinate tolerance, component ordering, empty geometries, and dimensional behavior.

Add focused tests at the native expression and GeoPandas layers, including tolerance boundaries, reordered geometries, scalar and series operands, index alignment, nulls, empty geometries, geometry families, and dimensional edge cases. Add the public API documentation required for a new GeoPandas-compatible method.

### PR DESCRIPTION
**Pull request by jiayuasu**
## Did you read the Contributor Guide?

- Yes, I have read the [Contributor Rules](https://sedona.apache.org/latest/community/rule/) and [Contributor Development Guide](https://sedona.apache.org/latest/community/develop/)

## Is this PR related to a ticket?

- Yes, and the PR name follows the format `[GH-XXX] my subject`. Closes #3156. Closes #3157. Closes #3158.
- Part of #2230.

## What changes were proposed in this PR?

- Implement distributed `GeoSeries.explode` with GeoPandas-compatible index modes, deterministic component ordering, geometry metadata preservation, and empty/null handling.
- Implement distributed `GeoSeries.get_coordinates` and its `GeoDataFrame` delegation, including optional Z/M ordinates and GeoPandas-compatible index modes.
- Share the `ST_Dump`/`ST_DumpPoints` plus `posexplode` row-expansion path between both expansion methods.
- Preserve every available GeoSeries index level through unary operations while keeping projected binary-operation index metadata in sync.
- Add the native `ST_EqualsExact` predicate with coordinate tolerance to Sedona common; expose it through Spark SQL, Flink SQL, and Snowflake BINARY/GEOMETRY/GEOGRAPHY functions; and add the Scala and Python DataFrame APIs.
- Implement distributed `GeoSeries.geom_equals_exact` and its `GeoDataFrame` delegation for scalar geometry and distributed Series operands, including duplicate-index and MultiIndex alignment behavior.
- Keep geometry processing in Spark without Python UDFs or driver-side geometry collection.

## How was this patch tested?

- Added GeoPandas parity tests covering geometry families, index modes, named and multi-level indexes, duplicate indexes, empty and null geometries, all-empty CRS preservation, output reuse, Z/M ordinates, and invalid inputs.
- Ran the consolidated focused GeoPandas suite: 41 tests passed.
- Ran the common predicate suite: 35 tests passed.
- Ran the focused Python SQL and DataFrame API suites: 5 tests passed.
- Ran the focused Flink predicate test: 1 test passed.
- Compiled the Snowflake and Snowflake tester modules, packaged the Snowflake module, verified all three generated DDL signatures, and checked the tolerance boundary directly.
- Built the Spark shaded package successfully with Spotless checks.
- Ran repository formatting and pre-commit checks.

## Did this PR include necessary documentation updates?

- Yes, I am adding new APIs. I am using the [current SNAPSHOT version number](https://github.com/apache/sedona/blob/99239524f17389fc4ae9548ea88756f8ea538bb9/pom.xml#L29) in `v1.9.1` format.
- Added public API docstrings and examples for all three GeoPandas methods.
- Added the `ST_EqualsExact` SQL references for Spark, Flink, and Snowflake.

### CODE REVIEW

File: python/sedona/spark/sql/st_predicates.py [hunk 1]
```diff
@@ -91,6 +91,29 @@ def ST_Equals(a: ColumnOrName, b: ColumnOrName) -> Column:
...
+    within a tolerance.
+
+    The comparison uses x and y coordinates. Z and M coordinates are ignored.
+
+    :param a: One geometry column to check.
+    :type a: ColumnOrName
+    :param b: Other geometry column to check.
+    :type b: ColumnOrName
+    :param tolerance: Maximum distance allowed between corresponding coordinates.
+    :type tolerance: ColumnOrName or float
```
[the remark this answers is not in the document]
Comment (jiayuasu) [reply python/sedona/spark/sql/st_predicates.py]: Fixed in 15cc146da02: the tolerance type documentation now includes int, matching the function signature.

File: docs/api/sql/Predicates/ST_EqualsExact.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+ -->
+
+# ST_EqualsExact
+
+Introduction: Return true if A and B have the same structure and their corresponding coordinates are equal within a tolerance.
+
+Unlike `ST_Equals`, this predicate requires geometry types, component order, ring order, and vertex order to match. The tolerance is the maximum distance allowed between each pair of corresponding coordinates. The comparison uses x and y coordinates and ignores z and m coordinates.
```
[the remark this answers is not in the document]
Comment (jiayuasu) [reply docs/api/sql/Predicates/ST_EqualsExact.md]: Fixed in 15cc146da02: changed the introduction to use Returns true and capitalized Z and M.
